sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -280,6 +280,29 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
   }
 
   /**
+   * Extracts the [[StructField]] with the given name recursively.
+   *
+   * @throws IllegalArgumentException if the parent field's type is not StructType
+   */
+  def getFieldRecursively(name: String): StructField = {
```

Comment:
I think there's another way to solve this problem, maybe generate the final structType in FileSourceStrategy better.I'll try it and give another pr later.